### PR TITLE
[mig-manager] move library/nvidia-smi discovery logic to Go

### PR DESCRIPTION
**Pull request by cdesiniotis**

### CODE REVIEW

File: cmd/nvidia-mig-manager/main.go [hunk 1]
```diff
@@ -70,6 +71,8 @@ var (
 	driverRootCtrPath string
 	devRoot           string
 	devRootCtrPath    string
+	driverLibraryPath string
+	nvidiaSMIPath     string
```
Comment (elezar) [inline cmd/nvidia-mig-manager/main.go:75]: I understand that we've been using globals for the other values here, but let's rather return these from the function where we extract them.

File: cmd/nvidia-mig-manager/main.go [hunk 2]
```diff
@@ -338,7 +384,7 @@ func runScript(migConfigValue string) error {
 		"-p", defaultGPUClientsNamespaceFlag,
 	}
 	if cdiEnabledFlag {
-		args = append(args, "-e", "-t", driverRoot, "-a", driverRootCtrPath, "-b", devRoot, "-j", devRootCtrPath)
+		args = append(args, "-e", "-t", driverRoot, "-a", driverRootCtrPath, "-b", devRoot, "-j", devRootCtrPath, "-l", driverLibraryPath, "-q", nvidiaSMIPath)
```
Comment (elezar) [inline cmd/nvidia-mig-manager/main.go:387]: the new parameters should be arguments to this function.

File: cmd/nvidia-mig-manager/main.go [hunk 3]
```diff
@@ -298,6 +306,44 @@ func start(c *cli.Context) error {
...
+// it is required to run nvidia-smi to create the nvidia-cap* device nodes.
+// If driverRoot != devRoot, we must discover the paths to libnvidia-ml.so.1 and
+// nvidia-smi in order to run nvidia-smi. We discover the paths here once and
+// pass these as arguments to reconfigure-mig.sh
+//
+// Currently, driverRoot != devRoot only when devRoot='/'. Since mig-manager
+// has rw access to the host rootFS (at hostRootMountFlag), reconfigure-mig.sh
+// will first chroot into the host rootFS before invoking nvidia-smi, so the
+// device nodes get created at '/dev' on the host.
+func setupRequiredForCDI() error {
```
Comment (elezar) [inline cmd/nvidia-mig-manager/main.go:322]: This name seems "off" we're not setting anythinng up.

Could we change this to:
```suggestion
func getPathsForCDI() (string, string, error) {
```
I would even go so far as to say that we want: `getDriverPaths() (string, string, error)` that actually contains the path logic and we can call this from `getPathsForCDI`.

File: deployments/container/reconfigure-mig.sh [hunk 4]
```diff
@@ -111,7 +115,13 @@ while getopts "hrden:f:c:m:i:o:g:k:p:t:a:b:j:" opt; do
...
       DEV_ROOT_CTR_PATH=${OPTARG}
       ;;
-    \? ) echo "Usage: ${0} -n <node> -f <config-file> -c <selected-config> -p <default-gpu-clients-namespace> [-e -t <driver-root> -a <driver-root-ctr-path> -b <dev-root> -j <dev-root-ctr-path>] [ -m <host-root-mount> -i <host-nvidia-dir> -o <host-mig-manager-state-file> -g <host-gpu-client-services> -k <host-kubelet-service> -r -s ]"
+    l ) # process option l
+      DRIVER_LIBRARY_PATH=${OPTARG}
+      ;;
+    q ) # process option q
+      NVIDIA_SMI_PATH=${OPTARG}
+      ;;
+    \? ) echo "Usage: ${0} -n <node> -f <config-file> -c <selected-config> -p <default-gpu-clients-namespace> [-e -t <driver-root> -a <driver-root-ctr-path> -b <dev-root> -j <dev-root-ctr-path> -l <driver-library-path> -q <nvidia-smi-path> ] [ -m <host-root-mount> -i <host-nvidia-dir> -o <host-mig-manager-state-file> -g <host-gpu-client-services> -k <host-kubelet-service> -r -s ]"
```
Comment (elezar) [inline deployments/container/reconfigure-mig.sh:124]: Any reason why we don't call `usage` and exit here? (like we do for `h`)

**Review by elezar — CHANGES_REQUESTED** [review]
Thanks @cdesiniotis.

It would be better if we move away from global variables here -- especially for the new code.

File: cmd/nvidia-mig-manager/main.go (hunk 1 above)
Comment (cdesiniotis) [reply cmd/nvidia-mig-manager/main.go]: Updated.

File: cmd/nvidia-mig-manager/main.go (hunk 2 above)
Comment (cdesiniotis) [reply cmd/nvidia-mig-manager/main.go:383]: Done.

File: cmd/nvidia-mig-manager/main.go (hunk 3 above)
Comment (cdesiniotis) [reply cmd/nvidia-mig-manager/main.go]: Agreed. Updated to `getPathsForCDI`.

File: deployments/container/reconfigure-mig.sh (hunk 4 above)
Comment (cdesiniotis) [reply deployments/container/reconfigure-mig.sh:124]: I am not aware of any reason not to. I can update this to call `usage` and `exit` in a follow-up.

**Review by elezar — APPROVED** [review]
Looks good now.

**Pull request merged**